### PR TITLE
fix possible missing gc-rootedness in REPL test

### PR DESCRIPTION
**Pull request by vtjnash**
Previously, we might see the gc finalize the REPL object,
and thus the STDOUT stream reader, too early.
This would leave the STDOUT stream writer with a broken pipe,
resulting in EPIPE errors during the test.

Issue was observed during a Travis CI run for a PR (https://travis-ci.org/JuliaLang/julia/jobs/262021492), and has not been reproduced locally to confirm.

Also took this opportunity to update a few of the constructs
used in the test to more modern Julia APIs.

### CODE REVIEW

File: test/repl.jl [hunk 1]
```diff
@@ -32,29 +41,24 @@ ccall(:jl_exit_on_sigint, Void, (Cint,), 0)
 # this should make sure nothing crashes without depending on how exactly the control
 # characters are being used.
 if !Sys.iswindows() || Sys.windows_version() >= Sys.WINDOWS_VISTA_VER
-    stdin_write, stdout_read, stderr_read, repl = fake_repl()
-
+    fake_repl() do stdin_write, stdout_read, repl
```
Comment (tkelman) [inline test/repl.jl]: why not replace the vista check with this line so the indentation is still right without having to change the block

File: test/repl.jl [hunk 2]
```diff
@@ -493,11 +498,12 @@ for prompt = ["TestΠ", () -> randstring(rand(1:10))]
 
     # Try entering search mode while in custom repl mode
     LineEdit.enter_search(s, custom_histp, true)
+    end
 end
 
 # Test removal of prompt in bracket pasting
 begin
-    stdin_write, stdout_read, stderr_read, repl = fake_repl()
+    fake_repl() do stdin_write, stdout_read, repl
```
Comment (tkelman) [inline test/repl.jl]: can replace the begin with this?

Comment (vtjnash) [inline test/repl.jl]: I could, but I'm either going to change all of them or none of them. There's no sense in doing half of them.

Comment (tkelman) [inline test/repl.jl]: all seems preferable to none, the diff can always be looked at with `?w=1`

Comment (vtjnash) [inline test/repl.jl]: ok, will merge when CI passes

**Review by tkelman — APPROVED** [review]

**Pull request merged**